pkg/cmd/provisioning/azure/create_managed_identities.go

```diff
@@ -0,0 +1,467 @@
+	default:
+		return nil, fmt.Errorf("found %q role definitions for %q, expected one", len(roleDefinitions), roleName)
+	}
+}
+
+// assignRoleToManagedIdentity assigns the Azure role with roleName within the provided subscriptionID to the
+// managed identity identified by managedIdentityPrincipalID within the provided scope.
+//
+// Scope is a string such as /subscriptions/<subscriptionID> which represents anything within the subscription.
+// This scope can be restricted within a resourceGroup such as /subscriptions/<subscriptionID>/resourceGroups/<resourceGroupName>.
+func assignRoleToManagedIdentity(client *azureclients.AzureClientWrapper, managedIdentityPrincipalID, roleName, scope, subscriptionID string) error {
+	targetRole, err := getRoleDefinition(client, roleName, subscriptionID)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to get role definition for role %s", roleName))
+	}
+
+	uuid, err := uuid.NewV4()
+	if err != nil {
+		return fmt.Errorf("failed to generate UUID for user-assigned identity role assignment: %v", err)
+	}
+	// Create a unique name for the role assignment
+	roleAssignmentName := uuid.String()
+	var rawResponse *http.Response
+	// Role assignment can fail due to a replication delay after creating the user-assigned managed identity
+	// Try up to 20 times with a 10 second delay each attempt
+	for i := 0; i < 20; i++ {
+		ctxWithResp := runtime.WithCaptureResponse(context.Background(), &rawResponse)
+		// armauthorization.RoleAssignmentsClientCreateResponse stomped because we don't need any information from
+		// the response however, ccoctl could potentially output the role assignment's ID similarly to other info
+		// which is log.Print'd
+		_, err = client.RoleAssignmentClient.Create(
+			ctxWithResp,
+			scope,
+			roleAssignmentName,
+			armauthorization.RoleAssignmentCreateParameters{
+				Properties: &armauthorization.RoleAssignmentProperties{
+					PrincipalID:      to.Ptr(managedIdentityPrincipalID),
+					RoleDefinitionID: targetRole.ID,
+				},
+			},
+			&armauthorization.RoleAssignmentsClientCreateOptions{},
+		)
+		if err != nil {
+			var respErr *azcore.ResponseError
+			if errors.As(err, &respErr) {
+				switch respErr.ErrorCode {
+				case "PrincipalNotFound":
+					// Replication delay. The identity we just created can't be found yet so we need to retry.
+					// TODO: Would it be better to display a message like this while we retry or would it be
+					//       acceptable to output a log line only once we've finished assigning the role?
+					if i == 19 {
+						log.Fatal("Timed out assigning role to user-assigned managed identity, this is most likely due to a replication delay following creation of the user-assigned managed identity, please retry")
+						break
+					} else {
+						log.Printf("Unable to assign role to user-assigned managed identity, retrying...")
+						time.Sleep(10 * time.Second)
+						continue
+					}
+				case "RoleAssignmentExists":
```

Comment:
Code review: double check that `"RoleAssignmentExists"` means that the role was already assigned and does not indicate a collision in our chosen uuid.

Reply:
Confirmed.